extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/vpc_flow_log_unmarshaler.go

```diff
@@ -0,0 +1,407 @@
+		logger:     logger,
+	}, nil
+}
+
+func (v *vpcFlowLogUnmarshaler) UnmarshalLogs(content []byte) (plog.Logs, error) {
+	var errGzipReader error
+	gzipReader, ok := v.gzipPool.Get().(*gzip.Reader)
+	if !ok {
+		gzipReader, errGzipReader = gzip.NewReader(bytes.NewReader(content))
+	} else {
+		errGzipReader = gzipReader.Reset(bytes.NewReader(content))
+	}
+	if errGzipReader != nil {
+		if gzipReader != nil {
+			v.gzipPool.Put(gzipReader)
+		}
+		return plog.Logs{}, fmt.Errorf("failed to decompress content: %w", errGzipReader)
+	}
+	defer func() {
+		_ = gzipReader.Close()
+		v.gzipPool.Put(gzipReader)
+	}()
+
+	switch v.fileFormat {
+	case fileFormatPlainText:
+		return v.unmarshalPlainTextLogs(gzipReader)
+	case fileFormatParquet:
+		// TODO
+		return plog.Logs{}, errors.New("still needs to be implemented")
+	default:
+		// not possible, prevent by NewVPCFlowLogUnmarshaler
+		return plog.Logs{}, nil
+	}
+}
+
+type resourceKey struct {
+	accountID string
+	region    string
+}
+
+func (v *vpcFlowLogUnmarshaler) unmarshalPlainTextLogs(reader io.Reader) (plog.Logs, error) {
+	scanner := bufio.NewScanner(reader)
+
+	var fields []string
+	if scanner.Scan() {
+		firstLine := scanner.Text()
+		fields = strings.Split(firstLine, " ")
+	}
+
+	scopeLogsByResource := map[resourceKey]plog.ScopeLogs{}
+	for scanner.Scan() {
+		line := scanner.Text()
+		if err := v.addToLogs(scopeLogsByResource, fields, line); err != nil {
+			return plog.Logs{}, err
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return plog.Logs{}, fmt.Errorf("error reading log line: %w", err)
+	}
+
+	return v.createLogs(scopeLogsByResource), nil
+}
+
+// createLogs based on the scopeLogsByResource map
+func (v *vpcFlowLogUnmarshaler) createLogs(scopeLogsByResource map[resourceKey]plog.ScopeLogs) plog.Logs {
+	logs := plog.NewLogs()
+
+	for key, scopeLogs := range scopeLogsByResource {
+		rl := logs.ResourceLogs().AppendEmpty()
+		attr := rl.Resource().Attributes()
+		attr.PutStr(conventions.AttributeCloudProvider, conventions.AttributeCloudProviderAWS)
+		if key.accountID != "" {
+			attr.PutStr(conventions.AttributeCloudAccountID, key.accountID)
+		}
+		if key.region != "" {
+			attr.PutStr(conventions.AttributeCloudRegion, key.region)
+		}
+		scopeLogs.MoveTo(rl.ScopeLogs().AppendEmpty())
+	}
+
+	return logs
+}
+
+// addToLogs parses the log line and creates
+// a new record log. The record log is added
+// to the scope logs of the resource identified
+// by the resourceKey created from the values.
+func (v *vpcFlowLogUnmarshaler) addToLogs(
+	scopeLogsByResource map[resourceKey]plog.ScopeLogs,
+	fields []string,
+	logLine string,
+) error {
+	// first line includes the fields
+	// TODO Replace with an iterator starting from go 1.24:
+	// https://pkg.go.dev/strings#FieldsSeq
+	nFields := len(fields)
+	nValues := strings.Count(logLine, " ") + 1
+	if nFields != nValues {
+		return fmt.Errorf("expect %d fields per log line, got log line with %d fields", nFields, nValues)
+	}
```

Comment:
For a followup:

Unless there's a bug in the parser, the number of fields should always match, so let's optimise for that. I suggest we remove this check here and count the fields as we're parsing to reduce overhead. Then at the very end we can return an error if the number of fields parsed does not match expectations.

Reply:
For sure, here is the issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/39043.